rfcs/unboxed-types.md

```diff
@@ -0,0 +1,1989 @@
+system for now, because it adds complication without anyone
+asking for it.
+
+Here are some complications introduced:
+
+* It is very convenient to reduce computing
+`la ⊓ lb` to a constraint `rep(la) = rep(lb)`. In order to support `value ⊓ word
+= immediate`, we would thus drop the `value_rep` representation, setting
+`rep(value) = word_rep`.
+
+* The `upper_bound` function we use in type inference then becomes impossible to
+define: neither `value` nor `word` is an upper bound for the other, and so
+`upper_bound(word_rep)` has no definition.  We thus have to design a different
+type inference algorithm than the one described above, likely tracking
+representation, gc-friendliness, and gc-ignorability separately. This would not
+be all that hard, but it's definitely more complicated than the algorithm
+proposed here.
+
+* We sometimes might infer `immediate` even when no function
+nearby has anything of layout `immediate`:
+
+    ```ocaml
+    val f1 : ('a : value). 'a -> ...
+    val f2 : ('a : word). 'a -> ...
+
+    let f3 x = (f1 x, f2 x)
+    ```
+
+    Here, we need to infer that the `x` in `f3` has layout
+    `immediate`, even though neither `f1` nor `f2` mentions
+    `immediate`. This might be unexpected by users.
+
+## Further discussion: mutability within unboxed records
+
+This proposal forbids mutability within an unboxed record (unless that
+record is itself contained within a boxed record). This decision is made to
+reduce the possibility of confusion and bugs, not because mutability within
+unboxed records is unsound. Here, we consider some reasons that push against the
+decision above and include some examples illustrating the challenges.
+
+Reasons we might want mutability within unboxed types:
+
+* The designer of a type may choose to have some fields be mutable while
+  other fields would not be. For example, our `#mut_point` type allows the
+  location of the point to change, while the identity of the point must
+  remain the same.
+
+* OCaml programmers expect to be able to update their mutable fields, and it
+  may cause confusion if a programmer cannot, say, update the `x` field of a
+  `#mut_point`.
+
+The reason we decided not to allow mutability within unboxed types is because of
+examples like this:
+
+```ocaml
+type mut_point = { mutable x : #int32; mutable y : #int32; identity : int }
+let () =
+  let p : #mut_point = #{ x = #4l; y = #5l; identity = 0 } in
+  let foo () = print_uint32 p.x in
+  foo ();
+  p.x <- #10l;
+  foo ()
+```
+
+Because the closure for `foo` captures that unboxed record `p` (by copying `p`),
+this code prints 4 twice, rather confusingly. Yet the current design does not
+fully rule such a confusion out, as we see here:
+
+```ocaml
+type point = { x : int; y : int }
+type t = { mutable pt : #point }
+
+let () =
+  let t : t = { pt = #{ x = 5; y = 6 } } in
+  let pt = t.pt in
+  let foo () = print_int pt.x in
+  foo ();
+  t.(.pt.x) <- 7;
+  foo ()
+```
```

Comment:
In our internal design debates, we eventually settled on a medium-volume notation: `t.(.pt.x) <- 7`, where the parentheses and extra dot are the extra volume. Originally, it was just `t.pt.x <- 7`, but that indeed seemed too confusing. Yet I sense from your comments that you think the design is not quite loud enough. Perhaps you're right.

Reply:
To clarify, there are two points in my comment that can be discussed separately:
* Maybe one can skip the artificially-loud notation altogether, by writing something like `t.pt <- #{ t.pt with x = 7 }` instead of `t.(.pt.x) <- 7` if it is made possible (after relaxing the tearing restriction: this is discussed in another comment). The compiler should be smart enough to compile it the same. Then all possible confusions about aliasing disappear. The notation is still loud, but not in way that has to be introduced artificially. (In my current understanding of this proposal, this is my preference compared to the next point.)
* _If_ nevertheless something like `t.pt.x <- 7` is kept, I pointed out that the confusing part (to me at least) is that the notation `let pt = t.pt` suggests an aliasing of the `x` fields that does not exist, so perhaps the latter operation is what deserves to be louder/clearer (e.g. `let pt = #{ t.pt }`).

Reply:
Subsumed by the discussion in the main thread, feel free to mark as resolved.